Add support for UDC for Apollo510 family

zeonchew opened the pull request:
This PR adds support for UDC for Ambiq Apollo510 family. The boards tested are:
- apollo510_evb

Followings are the USB samples tested: 
- cdc_acm 
- console-next
- mass
- hid-keyboard (with `CONFIG_PM_DEVICE=n`)
- hid-mouse (with `CONFIG_PM_DEVICE=n`)
- webusb-next

zeonchew committed:
boards: ambiq: apollo510: add USB nodes

Add USB node to apollo510 qualifier, and apollo510_evb board to enable
USB support on the SoC and its EVB.

Signed-off-by: Chew Zeh Yang <zeon.chew@ambiq.com>

zeonchew committed:
drivers: udc_ambiq: add support for apollo510

Added UDC support for Ambiq apollo510 family with USB HS and DMA support.

Signed-off-by: Chew Zeh Yang <zeon.chew@ambiq.com>

swift-tk commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -17,6 +17,7 @@
 #include <zephyr/logging/log.h>
 #include <zephyr/irq.h>
 #include "udc_common.h"
+#include "am_mcu_apollo.h"
```
Please use <soc.h> from now on.

zeonchew replied:
You are right! Didn't noticed that `am_mcu_apollo.h` is added into soc.h for all apollo series lately. 
Thanks for pointing out. Change has been made to remove the line (soc.h is included further above)

swift-tk commented:
One more suggestion, consider adding the controller to at least one USB examples/testcase for compile. i.e. Add apollo510_evb board to platform_allow and/or integration_platforms. 
Everything else LGTM.

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -542,10 +587,74 @@ static int udc_ambiq_init(const struct device *dev)
 	/* Release USB PHY reset */
 	am_hal_usb_disable_phy_reset_override();
+
+#if CONFIG_SOC_SERIES_APOLLO5X
```
Move to init_apollo5x()

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -542,10 +587,74 @@ static int udc_ambiq_init(const struct device *dev)
+	am_hal_usb_phy_clock_enable(priv->usb_handle, true, priv->usb_speed);
+#endif
+
+#if CONFIG_UDC_AMBIQ_DEB_ENABLE
```
Move to init_double_buffers()

zeonchew replied:
Hi @jfischer-no, moved to `init_double_buffers()` as suggested. Thanks.

zeonchew commented on drivers/usb/udc/Kconfig.ambiq:
```diff
@@ -33,4 +34,56 @@ config UDC_AMBIQ_MAX_QMESSAGES
+	  Packet Size. However this option cannot be chosen when synchronous
+	  endpoint is used.
+
+config UDC_AMBIQ_PIO_MODE
+	bool "PIO transfer mode"
+	help
+	  Select this config when cache coherency handling is to be
+	  avoided.
+endchoice
+
+config UDC_AMBIQ_HANDLE_CACHE
+	bool "Cache handling enable for USB non-control endpoint buffers"
+	default y
+	depends on CACHE_MANAGEMENT
+	depends on DCACHE
+	depends on !UDC_BUF_FORCE_NOCACHE
+	depends on UDC_AMBIQ_DMA0_MODE || UDC_AMBIQ_DMA1_MODE
+	help
+	  Disable this if non-cacheable memory is selected as default system memory
```
Hi @jfischer-no, 
The option to disable cache handling fully when they are not needed, saving CPU cycles used for the cache handling operations, as the `depends on` conditions above suggests.

zeonchew replied:
Hi @jfischer-no, I think I understand what you are envisioning for the cache handling now. I have removed the cache handling from this PR in the latest push.

zeonchew committed:
drivers: udc_ambiq: added support for double endpoint buffer

implements option to enable double endpoint buffer to improve throughtput
for non-control endpoints.

Signed-off-by: Chew Zeh Yang <zeon.chew@ambiq.com>

zeonchew commented on drivers/usb/udc/Kconfig.ambiq:
```diff
@@ -33,4 +34,56 @@ config UDC_AMBIQ_MAX_QMESSAGES
+	  Disable this if non-cacheable memory is selected as default system memory
+
+
+config UDC_AMBIQ_DEB_ENABLE
```
Hi @jfischer-no, I have split the Double Endpoint Buffer change to a separate commit. Please have a look again.

zeonchew commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -43,6 +43,12 @@ K_MSGQ_DEFINE(drv_msgq, sizeof(struct udc_ambiq_event), CONFIG_UDC_AMBIQ_MAX_QME
+#define XFER_LEN_CB_TYPE uint32_t
+#else
+#define XFER_LEN_CB_TYPE uint16_t
+#endif
```
Hi @jfischer-no, This part has been changed to use uint32_t as parameter.

zeonchew commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -473,11 +499,21 @@ static int udc_ambiq_disable(const struct device *dev)
+				   &int_status[3], &int_status[4]);
+	am_hal_usb_interrupt_service(priv->usb_handle, int_status[0], int_status[1], int_status[2],
+				     int_status[3], int_status[4]);
+#else
```
Hi @jfischer-no, I have addressed this using LISTIFY. Please help to review again. Thanks.

zeonchew commented:
Ping @jfischer-no, requested change has been made, please kindly help to review again.

AlessandroLuo commented:
@kartben it has been more than 2 weeks since we made changes as the request, however we cannot get in touch with @jfischer-no either here or discord, can you help on this? thank you!

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -107,6 +107,13 @@ static int udc_ambiq_tx(const struct device *dev, uint8_t ep, struct net_buf *bu
+#if defined(CONFIG_UDC_AMBIQ_HANDLE_CACHE)
+		/* Cache management if cache and DMA is enabled */
+		if ((ep != USB_CONTROL_EP_OUT) &&
+		    !buf_in_nocache((uintptr_t)buf->data, buf->size)) {
```
buf_in_nocache() looks like copied from some *_stm32 driver. Now we have multiple implementations of buf_in_nocache() in the tree, this one is not even properly prefixed and not properly implemented so that it could be optimized away if cache management is not enabled. Looks like all of them could be just mem_attr_check_buf(), but why the driver needs to check it here? What about "saving CPU cycles"?

zeonchew replied:
Hi @jfischer-no, understood on your comments. I agree that this part of code would need some improvement. I have removed the checking (as well as `CONFIG_UDC_AMBIQ_HANDLE_CACHE`) from this PR. Will study further on the cache handling and raise a new PR if any change regarding cache handling in `udc_ambiq` is to be proposed later on.

zeonchew committed:
west.yml: sync hal ambiq

sync latest hal to take in implementaiton of unified usb isr handling

Signed-off-by: Chew Zeh Yang <zeon.chew@ambiq.com>

zeonchew commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -579,6 +579,36 @@ static int init_apollo5x(const struct udc_ambiq_data *priv)
+							   AM_HAL_USB_OUT_DIR, true);
+		}
+		idx++;
+		mask >>= 1;
```
Hi @jfischer-no, yeah, it makes a lot of sense to use `find_lsb_set()`. I have changed the implementation to use `find_lsb_set()` instead. Thanks again for pointing this out.

zeonchew commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -579,6 +579,36 @@ static int init_apollo5x(const struct udc_ambiq_data *priv)
+	while (mask) {
+		if (mask & 0x1) {
+			am_hal_usb_enable_ep_double_buffer(priv->usb_handle, idx, AM_HAL_USB_IN_DIR,
+							   true);
```
Thanks for pointing out. Its fixed now.

zeonchew commented:
Hi @jfischer-no, ping. Please help to review again. Thank you.

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -470,14 +483,11 @@ static int udc_ambiq_disable(const struct device *dev)
 	return 0;
 }
 
+
```
Stray change.

zeonchew replied:
Hi @jfischer-no, Thanks for point this out, the stray change has been reverted in the correct commit now.

jfischer-no reviewed (approved):
LGTM

zeonchew commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -483,7 +483,6 @@ static int udc_ambiq_disable(const struct device *dev)
 	return 0;
 }
 
-
```
Hi @jfischer-no, Thanks for point this out, the stray change has been reverted in the correct commit now.

zeonchew commented:
> @zeonchew can you fix the stray change called out by @jfischer-no ? Once fixed happy to merge in.
Hi @dkalowsk, the stray change fixed in correct commit as pointed out. Thank you.
